python/mspasspy/db/database.py

```diff
@@ -1518,229 +1724,188 @@ def _check_required(self, collection='site', keys=['lat','lon','elev','starttime
-                        raise MsPASSError('Can not find the record with _id: {} in {} collection under {} mode.'.format(mspass_object['_id'], wf_collection, mode), 'Fatal')
-                    del mspass_object['_id']
-                    new_insertion = True
-
-            # 1. create the dict of metadata to be saved in wf
-            insert_dict = {}
-
-            # FIXME starttime will be automatically created in this function
-            self._sync_metadata_before_update(mspass_object)
-
-            copied_metadata = Metadata(mspass_object)
-
-            # clear all the aliases
-            update_metadata_def.clear_aliases(copied_metadata)
-
-            for k in copied_metadata:
-                if not str(copied_metadata[k]).strip():
-                    copied_metadata.erase(k)
-
+        if mode not in ['promiscuous', 'cautious', 'pedantic']:
+            raise MsPASSError(alg_name+': only promiscuous, cautious and pedantic are supported, but {} was requested.'.format(mode), 'Fatal')
+        if isinstance(mspass_object, TimeSeries):
+            save_schema = self.metadata_schema.TimeSeries
+        else:
+            save_schema = self.metadata_schema.Seismogram
+        if '_id' in mspass_object:
+            wfid=mspass_object['_id']
+        else:
+            raise MsPASSError(alg_name + ':  input data object is missing required waveform object id value (_id) - update is not possible without it','Fatal')
+        if collection:
+            wf_collection_name=collection
+        else:
+            # This returns a string that is the collection name for this atomic data type
+            # A weird construct
+            wf_collection_name=save_schema.collection('_id')
+        wf_collection=self.db[wf_collection_name]
+        # FIXME starttime will be automatically created in this function
+        self._sync_metadata_before_update(mspass_object)
+
+        # This method of Metadata returns a list of all
+        # attributes that were changed after creation of the
+        # object to which they are attached.
+        changed_key_list=mspass_object.modified()
+
+        copied_metadata = Metadata(mspass_object)
+
+        # clear all the aliases
+        #TODO  check for potential bug in handling clear_aliases
+        # and modified method - i.e. keys returned by modified may be
+        # aliases
+        save_schema.clear_aliases(copied_metadata)
+
+        # remove any defined items in exclude list
+        for k in exclude_keys:
+            if k in copied_metadata:
+                copied_metadata.erase(k)
+        # Now remove any readonly data
+        for k in copied_metadata.keys():
+            if save_schema.is_defined(k):
+                if save_schema.readonly(k):
+                    if k in changed_key_list:
+                        newkey='READONLYERROR_'+k
+                        copied_metadata.change_key(k,newkey)
+                        mspass_object.elog.log_error('Database.save_data',
```

Comment:
minor issue here, the elog message should be Database.update_metadata

Reply:
Good catch - cut and paste error